spec/models/questionnaire_response_spec.rb

```diff
@@ -0,0 +1,132 @@
+      now = Time.current
+      updated_questionnaire_response.device_updated_at = 20.minutes.ago
+      updated_questionnaire_response.content = {a: :b, c: :c}
+
+      Timecop.freeze(now) do
+        QuestionnaireResponse.merge(updated_questionnaire_response.attributes)
+      end
+
+      existing_questionnaire_response.reload
+      expect(existing_questionnaire_response.updated_at.to_i).to eq now.to_i
+      expect(existing_questionnaire_response.device_updated_at.to_i).to eq ten_minutes_ago.to_i
+      expect(existing_questionnaire_response.content).to eq({"a" => "a", "c" => "c"})
+      expect(QuestionnaireResponse.count).to eq 1
+    end
+
+    it "merges new keys in content when input record is equally up-to-date" do
+      ten_minutes_ago = 10.minutes.ago
+      existing_questionnaire_response = FactoryBot.create(
+        :questionnaire_response,
+        updated_at: ten_minutes_ago,
+        device_updated_at: ten_minutes_ago,
+        content: {a: :a}
+      )
+      updated_questionnaire_response = QuestionnaireResponse.find(existing_questionnaire_response.id)
+      now = Time.current
+      updated_questionnaire_response.device_updated_at = ten_minutes_ago
+      updated_questionnaire_response.content = {a: :b, c: :c}
+
+      Timecop.freeze(now) do
+        QuestionnaireResponse.merge(updated_questionnaire_response.attributes)
+      end
+
+      existing_questionnaire_response.reload
+      expect(existing_questionnaire_response.updated_at.to_i).to eq now.to_i
+      expect(existing_questionnaire_response.device_updated_at.to_i).to eq ten_minutes_ago.to_i
+      expect(existing_questionnaire_response.content).to eq({"a" => "b", "c" => "c"})
+      expect(QuestionnaireResponse.count).to eq 1
+    end
+
+    it "counts metrics for old merges" do
+      existing_questionnaire_response = FactoryBot.create(:questionnaire_response)
+      updated_questionnaire_response = QuestionnaireResponse.find(existing_questionnaire_response.id)
+
+      updated_questionnaire_response.device_updated_at = 10.minutes.ago
+
+      expect(Statsd.instance).to receive(:increment).with("merge.QuestionnaireResponse.updated")
+      QuestionnaireResponse.merge(updated_questionnaire_response.attributes)
+    end
+
+    it "counts metrics if the existing record device_updated_at is the same as the new one" do
+      timestamp = Time.zone.parse("March 1st 04:00:00 IST")
+      existing_questionnaire_response = FactoryBot.create(:questionnaire_response, device_updated_at: timestamp)
+      updated_questionnaire_response = QuestionnaireResponse.find(existing_questionnaire_response.id)
+
+      updated_questionnaire_response.device_updated_at = existing_questionnaire_response.device_updated_at
+
+      expect(Statsd.instance).to receive(:increment).with("merge.QuestionnaireResponse.updated")
+      QuestionnaireResponse.merge(updated_questionnaire_response.attributes)
+    end
+
+    it "acquires a lock before merging to avoid race conditions" do
```

Comment:
We tested the lock implementation [in this commit](https://github.com/simpledotorg/simple-server/pull/4683/commits/145948f1df69cd3636dabe38c372eacbbd89af44#diff-1df17d78ab6b0319fcc921b6c0bb36359d32d5d81c267ef4cae7cf22d07e64b5R126). Rspec tests run in a transaction and testing this required us to disable them. This left the DB in a broken state which was difficult to cleanup. Added this mock test instead.